### PR TITLE
URI Builder for ad types

### PR DESCRIPTION
**Pull request by lm-Lakhan-Balmiki**
URL Builder according to the Ad types.

### CODE REVIEW

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java [hunk 1]
```diff
@@ -17,11 +18,17 @@ public class MainActivity extends AppCompatActivity {
...
-    private static final String DEFAULT_URL_Aid_Pid =
-            "http://localhost:8000/openrtb2/video?test=1&tmax=3000&vw=1080&vh=1920&apid=abcd&apdom=%22test.prebid.com%22&apbndl=%22com.prebid.test%22&ip=%22192.168.1.2%22&vmimes=[%22video%2Fmp4%22%2C%20%22video%2Fog3%22]&wpid=178&waid=lemma-hb-adunit";
+    private static final String BASE_URL_AID_PID =
+            "http://sandbox.lemmatechnologies.com/infibid/v1/video/vast";
+
+    private static final String DEFAULT_VAST_TAG_PARAMS =
+            "?test=1&tmax=3000&vw=1920&vh=1080&apdom=%22test.prebid.com%22&apbndl=%22com.prebid.test%22&ip=%22192.168.1.2%22&vmimes=[%22video%2Fmp4%22%2C%20%22video%2Fog3%22]&wpid=178&waid=lemma-hb-adunit&gam_au=";
+
+    private static final String DEFAULT_URL_AID_PID_PARAMS =
+            "?test=1&tmax=3000&vw=1080&vh=1920&apid=abcd&apdom=%22test.prebid.com%22&apbndl=%22com.prebid.test%22&ip=%22192.168.1.2%22&vmimes=[%22video%2Fmp4%22%2C%20%22video%2Fog3%22]&wpid=178&waid=GAM_HB_demo";
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]: No need of this variable

**Review by lm-sushant-deshpande — CHANGES_REQUESTED** [review]

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java (hunk 1 above)
Comment (lm-Lakhan-Balmiki) [reply InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]: Resolved the issue.

**Review by lm-sushant-deshpande — CHANGES_REQUESTED** [review]
1. Use .idea/** in gitignore
2. Use https urls and remove network config file

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/GAMHBAdTagUrlBuilder.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    public void requestHBAdTargeting(TargetingCallback callback) {
+        // Hardcoded VAST URL for sandbox testing purposes.
+        String vastUrl = "https://sandbox.lemmatechnologies.com/infibid/v1/video/targeting";
+        // Use a LemmaVastAdTagUrlBuilder to construct the targeting URL.
+        LemmaVastAdTagUrlBuilder urlBuilder = new LemmaVastAdTagUrlBuilder(this.context, vastUrl, pubId, adUnitId);
+        String targetingUrl = urlBuilder.getUrl();
+
+        // Start a new thread to perform the network request asynchronously.
+        new Thread(() -> {
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/GAMHBAdTagUrlBuilder.java]: Remove thread

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/GAMHBAdTagUrlBuilder.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+                        try {
+                            // Parse the response and generate query parameters.
+                            queryParams = parser.parseAndGenerateQueryParams(responseData);
+                            callback.onSuccess(queryParams); // Invoke success callback with the parameters
+                        } catch (Exception e) {
+                            // Handle any errors that occur during parsing.
+                            Log.e(TAG, "Error parsing targeting response", e);
+                            callback.onFailure(e); // Invoke failure callback with the exception
+                        }
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/GAMHBAdTagUrlBuilder.java]: remove multiple try catch bolck

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java [hunk 4]
```diff
@@ -51,40 +61,71 @@ public void onItemSelected(AdapterView<?> parent, View view, int position, long
...
+                        // Build the URL with custom parameters and start the IMAPlayer activity.
+                        String url = urlBuilder.getUrl(custParam);
+                        Log.d(TAG, "Ad with GAM Ad Server URL: " + url);
+                        startIMAPlayerActivity(url);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        // Log the error and handle it, e.g., show a message to the user.
+                        Log.e(TAG, "Failed to fetch targeting data", e);
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]: call `startIMAPlayerActivity(url);`

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java [hunk 5]
```diff
@@ -51,40 +61,71 @@ public void onItemSelected(AdapterView<?> parent, View view, int position, long
...
+                    }
+                });
+
+            } else {
+                // Handle case where no ad server is selected or an unknown server is chosen.
+                String url = "";
+                startIMAPlayerActivity(url);
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]: Remove this

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java [hunk 6]
```diff
@@ -51,40 +61,71 @@ public void onItemSelected(AdapterView<?> parent, View view, int position, long
...
-                // Use the default URL_Aid_Pid if "No Ad Server" is selected
-                url = DEFAULT_URL_Aid_Pid.replace("178",pubId).replace("lemma-hb-adunit",adUnitId);
-                Log.d("Ad with No Ad Server", url);
-            } else if("GAM Ad Server".equals(selectedServer)) {
-                // Construct VAST URL based on the presence of GAM Ad Unit ID
-                    // Replace placeholder with actual GAM Ad Unit ID
-                    url = DEFAULT_VAST_TAG_URL.replace("/22192417927", "/" + gamAdUnitId);
-                    Log.d("GAM Ad with gam id",url);
+                // Build the VAST ad tag URL using LemmaVastAdTagUrlBuilder.
+                String vastUrl = "http://sandbox.lemmatechnologies.com/infibid/v1/video/vast";
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]: Move this url inside `LemmaVastAdTagUrlBuilder `

**Review by lm-sushant-deshpande — CHANGES_REQUESTED** [review]
1. Replace `sandbox.lemmatechnologies.com` with `lemmadigital.com`

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/GAMHBAdTagUrlBuilder.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+         * @param custParam The custom parameters extracted from the targeting response.
+         */
+        void onSuccess(String custParam);
+
+        /**
+         * Called when the targeting request fails.
+         *
+         * @param e The exception that caused the failure.
+         */
+        void onFailure(Exception e);
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/GAMHBAdTagUrlBuilder.java]: Use Error

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java [hunk 8]
```diff
@@ -51,40 +61,68 @@ public void onItemSelected(AdapterView<?> parent, View view, int position, long
...
+                Log.d(TAG, "Ad with No Ad Server URL: " + url);
+
+                // Start the IMAPlayer activity to play the video ad.
+                startIMAPlayerActivity(url);
+
+            } else if ("GAM Ad Server".equals(selectedServer)) {
+                // Build the GAM ad tag URL using GAMHBAdTagUrlBuilder and request targeting data.
+                GAMHBAdTagUrlBuilder urlBuilder = new GAMHBAdTagUrlBuilder(this, pubId, adUnitId, gamAdUnitId);
+
+                urlBuilder.requestHBAdTargeting(new GAMHBAdTagUrlBuilder.TargetingCallback() {
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]:
```suggestion
                urlBuilder.build(new GAMHBAdTagUrlBuilder.TargetingCallback() {
```

File: InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java [hunk 9]
```diff
@@ -51,40 +61,68 @@ public void onItemSelected(AdapterView<?> parent, View view, int position, long
...
-            } else if("GAM Ad Server".equals(selectedServer)) {
-                // Construct VAST URL based on the presence of GAM Ad Unit ID
-                    // Replace placeholder with actual GAM Ad Unit ID
-                    url = DEFAULT_VAST_TAG_URL.replace("/22192417927", "/" + gamAdUnitId);
-                    Log.d("GAM Ad with gam id",url);
-            }
+                // Build the VAST ad tag URL using LemmaVastAdTagUrlBuilder.
+                String vastUrl = "https://lemmadigital.com/infibid/v1/video/vast";
+                LemmaVastAdTagUrlBuilder urlBuilder = new LemmaVastAdTagUrlBuilder(this, vastUrl, pubId, adUnitId);
+                String url = urlBuilder.getUrl();
```
Comment (lm-sushant-deshpande) [inline InfiBidSampleApp/app/src/main/java/com/example/InfiBidSampleApp/MainActivity.java]:
```suggestion
                String url = urlBuilder.build();
```

**Review by lm-sushant-deshpande — CHANGES_REQUESTED** [review]

**Review by lm-sushant-deshpande — APPROVED** [review]

**Pull request merged**